cli/azd/pkg/pipeline/github_provider.go

```diff
@@ -627,18 +619,100 @@ func (p *GitHubCiProvider) configurePipeline(
+	msg := ""
+	var procErr error
+	ciSecrets, ciVariables := []string{}, []string{}
+	if len(options.Variables) > 0 || len(options.Secrets) > 0 {
+		msg = "Setting up project's secrets and variables to be used in the pipeline"
+		ciSecretsInstance, err := p.ghCli.ListSecrets(ctx, repoSlug)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get list of repository secrets: %w", err)
+		}
+		ciVariablesInstance, err := p.ghCli.ListVariables(ctx, repoSlug)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get list of repository variables: %w", err)
+		}
+		ciSecrets = ciSecretsInstance
+		ciVariables = ciVariablesInstance
+		p.console.ShowSpinner(ctx, msg, input.Step)
+	}
+
+	defer func() {
+		if msg != "" {
+			p.console.StopSpinner(ctx, msg, input.GetStepResultFormat(procErr))
+		}
+		if procErr == nil {
+			p.console.MessageUxItem(ctx, &ux.MultilineMessage{
+				Lines: []string{
+					"",
+					"GitHub Action secrets are now configured. You can view GitHub action secrets that were created at this link:",
+					output.WithLinkFormat("https://github.com/%s/settings/secrets/actions", repoSlug),
+					""},
+			})
+		}
+	}()
+
+	// combine the variables and secrets for O(1) lookup during clean up
+	combinedVariablesAndSecrets := make(map[string]string, len(options.Variables)+len(options.Secrets))
+	for _, value := range options.Variables {
+		combinedVariablesAndSecrets[value] = value
+	}
+	for _, value := range options.Secrets {
+		combinedVariablesAndSecrets[value] = value
+	}
+
+	// iterate the existing secrets on the pipeline and remove the ones matching the project's secrets or variables
+	for _, existingSecret := range ciSecrets {
+		if _, willBeUpdated := additionalSecrets[existingSecret]; willBeUpdated {
+			// if the secret will be updated, we don't need to delete it
+			continue
+		}
+		// only delete if the secret is defined in the project's secrets or variables (azure.yaml)
+		if _, exists := combinedVariablesAndSecrets[existingSecret]; exists {
+			deleteErr := p.ghCli.DeleteSecret(ctx, repoSlug, existingSecret)
+			if deleteErr != nil {
+				procErr = fmt.Errorf("failed deleting %s secret: %w", existingSecret, deleteErr)
+				return nil, procErr
+			}
+		}
+	}
+	// iterate the existing variables on the pipeline and remove the ones matching the project's secrets or variables
+	for _, existingVariable := range ciVariables {
+		if _, willBeUpdated := additionalVariables[existingVariable]; willBeUpdated {
+			// if the variable will be updated, we don't need to delete it
+			continue
+		}
+		// only delete if the variable is defined in the project's secrets or variables (azure.yaml)
+		if _, exists := combinedVariablesAndSecrets[existingVariable]; exists {
+			deleteErr := p.ghCli.DeleteVariable(ctx, repoSlug, existingVariable)
+			if deleteErr != nil {
+				procErr = fmt.Errorf("failed deleting %s variable: %w", existingVariable, deleteErr)
+				return nil, procErr
+			}
+		}
+	}
+
+	// set the new variables and secrets
```

Comment:
nit: variables and secrets may not be "new"

Reply:
there are new to the eyes of azd, as all previous ones have been deleted from the pipeline

Reply:
@vhvb1989 I'm confused about that. Isn't the following true? "// if the variable will be updated, we don't need to delete it"

Reply:
Yes, it is true.
For azd, the value for the variable is always new.. either to set/create a secret/variable or to update it